### ISSUE TITLE
Feature Request: Pass complex blocks as variables

### ISSUE THREAD

**Issue by AndresFPineros** [hashicorp/terraform#25668]
### Current Terraform Version
```
Version: 0.12.29
```

### Use-cases
I want to be able to pass complex nested blocks as variables to avoid variable defintion overhead. Let me explain with the following code extracted from https://www.terraform.io/docs/providers/kubernetes/r/pod.html:
```
resource "kubernetes_pod" "with_node_affinity" {
  metadata {
    name = "with-node-affinity"
  }

  spec {
    affinity {
      node_affinity {
        required_during_scheduling_ignored_during_execution {
          node_selector_term {
            match_expressions {
              key      = "kubernetes.io/e2e-az-name"
              operator = "In"
              values   = ["e2e-az1", "e2e-az2"]
            }
          }
        }

        preferred_during_scheduling_ignored_during_execution {
          weight = 1

          preference {
            match_expressions {
              key      = "another-node-label-key"
              operator = "In"
              values   = ["another-node-label-value"]
            }
          }
        }
      }
    }

    container {
      name  = "with-node-affinity"
      image = "k8s.gcr.io/pause:2.0"
    }
  }
}
```
Here, we can see there are many nested blocks which are optional. In my use case, I have no idea which of these blocks is going to be used in a specific instance of the deployment. This means someone could use `node_affinity` or not, and inside that block they could use `required_during_scheduling_ignored_during_execution` or `preferred_during_scheduling_ignored_during_execution`, and inside those blocks other variations.

It would be a "dynamic" rabbit-hole to handle all the possible scenarios in which a person can configure these nested blocks. 

What to be able to pass a variable like:
```
variable "node_affinity" {
  default = {
    required_during_scheduling_ignored_during_execution {
      node_selector_term {
        match_expressions {
          key      = "kubernetes.io/e2e-az-name"
          operator = "In"
          values   = ["e2e-az1", "e2e-az2"]
        }
      }
    }
  }
}
```
And inside the deployment use it like:
```
resource "kubernetes_pod" "with_node_affinity" {
  metadata {
    name = "with-node-affinity"
  }
  spec {
    affinity {
      node_affinity = var.node_affinity
    }
    container {
      name  = "with-node-affinity"
      image = "k8s.gcr.io/pause:2.0"
    }
  }
}
```
This way someone would configure the "node_affinity" block with any configuration they'd like without having to overthink the variable definition and type checking. I'm not sure if this contradicts the Terraform way of doing things, but it would be a nice to have. Allowing users to treat blocks like maps would make Terraform more dynamic and friendly.

**Comment by AndresFPineros** [comment]
I'd really like to see progress on this one. I think we'll have to start using the Terraform CDK to achieve this degree of flexibility as an alternative solution. Still, I don't understand why blocks can't be just passed as maps, it would be nice if there was an explanation of the rationale behind because it seems like deliberately removing flexibility without a benefit in return.

I understand that a contract wants to be enforced to ensure that random stuff or malicious variables won't be passed to the TF logic, but you could just follow this basic rule:
- If someone sets the block equal to a map, then allow the flexibility of setting anything in there. This is a risk they'll take.
- If someone defines the block with "dynamic", then enforce the contract.

This way people can choose between flexibility or security.

Also, I'd like to hear some feedback on the idea of moving to CDK to solve this issue.

Thanks :)

**Comment by lllamnyp** [comment]
Yes please! The lack of this feature forces module developers to replicate the entire API of a provider within their code or greatly constrain users to a subset of allowed parameters.

**Comment by davidlbyrne** [comment]
I can't believe you cannot pass a complex object as a var and just assign it as a single value to a block. Please tell me I'm the idiot and there is a better way to do this I just don't see it?  

I'm implementing a module for a waf and there may be 1 rule or maybe 20 rules. The way the statement logic for the waf rules are constructed uses a potentially deeply nested objects/map. If I want the consumer of my module to be able to define their own rules and pass them as a map of objects map(any) then I will have to implement every possible key, and child, and child... as dynamic block. 
For reference : https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/wafv2_web_acl.html#statement-block. 

example: 
```terraform 

variable "rules" {
  type  =  map(any)
  default = {
    block-non-usa-traffic = {
      name     = "block-non-usa-traffic"
      priority = 0
      action = {
        block = {
          custom_response = {
            response_code = 302
            response_header = {
              name  = "Location"
              value = "http://blocked.fartsmeller.com"
            }
          }
      } }
      statement = {
        not_statement = {
          statement = {
            geo_match_statement = {
              country_codes = [
                "US",
              ]
            }
          }
        }
      }
      visibility_config = {
        cloudwatch_metrics_enabled = true
        metric_name                = "block-non-usa-traffic"
        sampled_requests_enabled   = true
      }
    },  
```
The implementation is disgusting: 

```terraform 
resource "aws_wafv2_web_acl" "common" {
  name  = local.name_prefix
  scope = "REGIONAL"

  default_action {
    allow {}
  }
  dynamic "rule" {
    for_each = var.rules
    content {
      name     = "${local.name_prefix}-${rule.value.name}"
      priority = rule.value.priority
     # statement = rule.value["statement"]  ## <--- this will not work :( so you have to do something like this: ... 
     dynamic "statement" { 
       for_each = rule.value.statement
       content { 
         dynamic "not_statement" {
           for_each = contains(keys(rule.value.statement), "not_statement' )) == true ? rule.value.statement.not_statement :[]
             dynamic "geo_match_statement" {
               for_each = ...
               ...
              }
             ... 
             ...
            }
         }
       }
      action {
        dynamic "block" { 
        foreach = contains(keys(rule.value.action), "block") == true ? [1] : []
          custom_response {
            response_code = action.block.response_code
            response_header {
              name  = action.block.response_header.name
              value = action.block.response_header.value
            }
          }
        }
      }
```
When on the other hand you could just do this:
```terraform 

resource "aws_wafv2_web_acl" "common" {
  name  = local.name_prefix
  scope = "REGIONAL"

  default_action {
    allow {}
  }
  dynamic "rule" {
    for_each = var.rules
    content {
      name     = "${local.name_prefix}-${rule.value.name}"
      priority = rule.value.priority
     statement = rule.value["statement"] 
   }
}
 
     
```
Please tell me I'm the idiot and there is a better way to do this I just don't see it?

**Comment by eric-adams-petdesk** [comment]
I just ran into @davidlbyrne 's exact use-case. When terraform hits these kinds of deeply nested, polymorphic configuration structures, it becomes a rubber mallet where you needed a scalpel. 

This seems fixable, so it is disappointing that the team has not acknowledged the issue or suggested workarounds over 5+ years.

**Comment by apparentlymart** [comment]
FWIW, this idea of "blocks" is constrained as it is -- including the strange requirement to use `dynamic` blocks when declaring them dynamically -- for annoying legacy reasons.

Resource type implementations using the legacy Terraform Plugin SDK -- which still includes a lot of resource types in the `hashicorp/aws` provider because it's a very large, old provider -- tend to crash or return very strange error messages if Terraform tries to send them anything that doesn't follow the rigid block structure described in the schema, because the SDK and those providers were written at a time when Terraform's language was even less flexible than it is today and so they had never needed to deal with anything dynamic in those contexts before.

The `dynamic` block mechanism is carefully designed to carefully skirt around all of those legacy constraints and be _as dynamic as possible_ within what those legacy SDK implementations will tolerate without crashing or behaving in very strange ways.

Newer providers written using the modern [Terraform plugin framework](https://github.com/hashicorp/terraform-plugin-framework) tend to avoid these constraints by just not using "blocks" at all, and using the "attribute" syntax exclusively instead. That acts as a signal to Terraform that the provider is willing to accept all of the various things that a dynamic expression reference can produce, including null values and values that won't be known until the apply phase.

I would personally love it if these constraints did not exist, but the already-published providers are what they are and have the constraints they have. 😖 Hopefully over time the provider will gradually migrate to a more modern API design that would be more flexible, but it's a huge codebase so that can't happen all at once.

(I am not speaking on behalf of the Terraform team. I'm sharing this just in case it's useful context for why this is designed the way it is, and why it's challenging to improve any further.)
